src/transformers/trainer_tf.py

```diff
@@ -0,0 +1,509 @@
+
+                    if step == 1:
+                        with self.train_writer.as_default():
+                            tf.summary.trace_export(name="training", step=step, profiler_outdir=self.log_path)
+
+                    if step % 10 == 0:
+                        logger.info("Epoch {} Step {} Loss {:.4f} Train Accuracy {:.4f}".format(epoch, step.numpy(), total_loss.numpy() / num_batches, self.train_acc_metric.result()))
+
+                    if step % 100 == 0:
+                        ckpt_save_path = self.model.ckpt_manager.save()
+                        logger.info("Saving checkpoint for step {} at {}".format(step, ckpt_save_path))
+
+                    if step % self.train_steps == 0:
+                        step += 1
+                        break
+
+                    step += 1
+
+                train_loss = total_loss / num_batches
+
+                self._evaluate_during_training()
+
+                logger.info("Epoch {} Step {} Train Loss {:.4f} Train Accuracy {:.4f}".format(epoch, step.numpy() - 1, train_loss.numpy(), self.train_acc_metric.result()))
+                logger.info("Epoch {} Validation Loss {:.4f} Validation Accuracy {:.4f}".format(epoch, self.test_loss_metric.result(), self.test_acc_metric.result()))
+
+            if epoch != self.epochs:
+                self.train_acc_metric.reset_states()
+                self.test_acc_metric.reset_states()
+
+    @abstractmethod
+    def _distributed_test_step(self, dist_inputs):
+        """
+        Method that represents a custom test step in distributed mode
+        Args:
+          dist_inputs: the features batch of the test data
+        """
+        pass
+
+    @abstractmethod
+    def _distributed_train_step(self, dist_inputs):
+        """
+        Method that represents a custom training step in distributed mode.
+        Args:
+          dist_inputs: the features batch of the training data
+        """
+        pass
+
+    @abstractmethod
+    def evaluate(self):
+        """
+        Evaluate the model over the test dataset and print a report.
+        """
+        pass
+
+    def save_model(self, save_path):
+        """
+        Save the pretrained model and create a Tensorflow saved model.
+        Args:
+          save_path: directory path where the pretrained model and
+            Tensorflow saved model will be saved
+        """
+        logger.info("Saving model in {}".format(save_path))
+
+        path = os.path.join(save_path, "saved_model")
+
+        os.makedirs(path, exist_ok=True)
+        self.model.save_pretrained(save_path)
+        tf.saved_model.save(self.model, path)
+
+
+class TFTrainerForSequenceClassification(TFTrainer):
+    def __init__(self, train_examples=None, dev_examples=None, test_examples=None, **config):
+        model_config = config.pop("model_config", None)
+        data_processor_config = config.pop("data_processor_config", None)
```

Comment:
Maybe default to `{}` instead of `None`?

Reply:
Yes, I should also do a check: if both exists then it is ok, else fail.